File: Habitica/src/main/java/com/habitrpg/android/habitica/utils/TaskListDeserializer.java
package com.habitrpg.android.habitica.utils;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonArray;
import com.habitrpg.android.habitica.models.tasks.Task;
import com.habitrpg.android.habitica.models.tasks.TaskList;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by viirus on 09/12/15.
 */
public class TaskListDeserializer implements JsonDeserializer<TaskList> {

    private List<Integer> getIntListFromJsonArray(JsonArray jsonArray) {
        List<Integer> intList = new ArrayList<>();

        for (int i = 0; i < jsonArray.size(); i++) {
            intList.add(jsonArray.get(i).getAsInt());
        }

        return intList;
    }

    private void getMonthlyDays(JsonElement e, Task task) {
        JsonArray weeksOfMonth = e.getAsJsonObject().getAsJsonArray("weeksOfMonth");
        if (weeksOfMonth != null && weeksOfMonth.size() > 0) {
            task.setWeeksOfMonth(getIntListFromJsonArray(weeksOfMonth));
        }

        JsonArray daysOfMonth = e.getAsJsonObject().getAsJsonArray("daysOfMonth");
        if (weeksOfMonth != null && weeksOfMonth.size() > 0) {
            task.setDaysOfMonth(getIntListFromJsonArray(daysOfMonth));
        }
    }

    public TaskList deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext ctx) {
        TaskList tasks = new TaskList();
        Map<String, Task> taskMap = new HashMap<>();

        for (JsonElement e : json.getAsJsonArray()) {
            try {
                Task task = ctx.deserialize(e, Task.class);

                // Work around since Realm does not support Arrays of ints
                getMonthlyDays(e, task);

                //Workaround, since gson doesn't call setter methods
                task.setId(task.getId());
                taskMap.put(task.getId(), task);
            } catch (ClassCastException ignored) {

            }
        }

        tasks.tasks = taskMap;

        return tasks;
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/utils/TaskListDeserializer.kt
package com.habitrpg.android.habitica.utils

import com.google.firebase.perf.FirebasePerformance
import com.google.gson.JsonArray
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.habitrpg.android.habitica.models.Tag
import com.habitrpg.android.habitica.models.tasks.Task
import com.habitrpg.android.habitica.models.tasks.TaskList
import java.lang.reflect.Type

class TaskListDeserializer : JsonDeserializer<TaskList> {
    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        ctx: JsonDeserializationContext,
    ): TaskList {
        val tasks = TaskList()
        val taskMap = HashMap<String, Task>()
        val deserializeTrace =
            try {
                FirebasePerformance.getInstance().newTrace("TaskListDeserialize")
            } catch (ignored: IllegalStateException) {
                // Firebase isn't initialized outside a running app process (e.g. unit tests)
                null
            }
        deserializeTrace?.start()

        for (e in json.asJsonArray) {
            try {
                val obj = e as? JsonObject
                if (obj != null) {
                    val task = ctx.deserialize<Task>(obj, Task::class.java)
                    task.tags?.addAll(handleTags(obj.getAsJsonArray("tags"), ctx))
                    task.id?.let { taskMap[it] = task }
                }
            } catch (ignored: ClassCastException) {
            } catch (ignored: java.lang.UnsupportedOperationException) {
            }
        }

        tasks.tasks = taskMap
        deserializeTrace?.stop()
        return tasks
    }

    // Raw tag-id entries become thin (id-only) placeholders; TaskRepositoryImpl resolves them against local tags after parsing.
    private fun handleTags(
        json: JsonArray?,
        context: JsonDeserializationContext,
    ): List<Tag> {
        val tags = mutableListOf<Tag>()
        for (tagElement in json ?: listOf<JsonElement>()) {
            if (tagElement.isJsonObject) {
                tags.add(context.deserialize(tagElement, Tag::class.java))
            } else {
                try {
                    val tagId = tagElement.asString
                    if (tags.none { it.id == tagId }) {
                        tags.add(Tag().apply { id = tagId })
                    }
                } catch (ignored: UnsupportedOperationException) {
                }
            }
        }
        return tags
    }
}
